### ISSUE TITLE
LatestMilestoneTracker error while analyzing the milestone candidates

### ISSUE THREAD

**Issue by georgmittendorfer** [iotaledger/iri#1477]
### Bug description

Exception in log file from LatestMilestoneTracker: `MilestoneException: unexpected error while analyzing the milestone candidates`

### IRI version

Patched IRI 1.7.0-DEV

### Steps To Reproduce

I don't know.

### Errors

Reason: java.lang.IndexOutOfBoundsException in MilestoneServiceImpl:

```
06/02 04:43:55.702 [Latest Milestone Tracker] ERROR LatestMilestoneTrackerImpl:465 - error while analyzing the milestone candidates
com.iota.iri.service.milestone.MilestoneException: unexpected error while analyzing the transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:170) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.analyzeMilestoneCandidates(LatestMilestoneTrackerImpl.java:347) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.latestMilestoneTrackerThread(LatestMilestoneTrackerImpl.java:274) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at com.iota.iri.utils.thread.BoundedScheduledExecutorService.lambda$wrapTask$6(BoundedScheduledExecutorService.java:823) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_172]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[na:1.8.0_172]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_172]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[na:1.8.0_172]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_172]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_172]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_172]
Caused by: com.iota.iri.service.milestone.MilestoneException: unexpected error while analyzing the transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:222) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:168) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        ... 10 common frames omitted
Caused by: com.iota.iri.service.milestone.MilestoneException: error while checking milestone status of transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.MilestoneServiceImpl.validateMilestone(MilestoneServiceImpl.java:228) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:195) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        ... 11 common frames omitted
Caused by: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
        at java.util.LinkedList.checkElementIndex(LinkedList.java:555) ~[na:1.8.0_172]
        at java.util.LinkedList.get(LinkedList.java:476) ~[na:1.8.0_172]
        at com.iota.iri.service.milestone.impl.MilestoneServiceImpl.validateMilestone(MilestoneServiceImpl.java:178) ~[iri-1.7.0-DEV-6bde715.jar:1.7.0-DEV-6bde715]
        ... 12 common frames omitted
```
Attention: The line numbers of the stack trace might not be accurate as I did run a patched version of IRI.

Problem happens in LatestMilestoneTrackerImpl lines 166-172:

```

166: public boolean processMilestoneCandidate(Hash transactionHash) throws MilestoneException {
167:       try {
168:           return processMilestoneCandidate(TransactionViewModel.fromHash(tangle, transactionHash));
169:       } catch (Exception e) {
170:           throw new MilestoneException("unexpected error while analyzing the transaction " + transactionHash, e);
171:       }
172:   }
```

Line 177 and 178 of MilestoneServiceImpl:
```
177: final TransactionViewModel siblingsTx =
178:                                bundleTransactionViewModels.get(coordinatorSecurityLevel);
```

**Comment by georgmittendorfer** [comment]
TX that caused the exception: https://thetangle.org/transaction/XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999

**Comment by georgmittendorfer** [comment]
For reference. Same stacktrace with 1.7.0:

```06/04 14:20:50.273 [Latest Milestone Tracker] ERROR c.i.i.s.m.i.LatestMilestoneTrackerImpl - error while analyzing the milestone candidates
com.iota.iri.service.milestone.MilestoneException: unexpected error while analyzing the transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:189) ~[iri-1.7.0.jar:na]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.analyzeMilestoneCandidates(LatestMilestoneTrackerImpl.java:351) ~[iri-1.7.0.jar:na]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.latestMilestoneTrackerThread(LatestMilestoneTrackerImpl.java:287) ~[iri-1.7.0.jar:na]
        at com.iota.iri.utils.thread.BoundedScheduledExecutorService.lambda$wrapTask$6(BoundedScheduledExecutorService.java:749) ~[iri-1.7.0.jar:na]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_172]
        at java.util.concurrent.FutureTask.runAndReset(FutureTask.java:308) ~[na:1.8.0_172]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$301(ScheduledThreadPoolExecutor.java:180) ~[na:1.8.0_172]
        at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:294) ~[na:1.8.0_172]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149) ~[na:1.8.0_172]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624) ~[na:1.8.0_172]
        at java.lang.Thread.run(Thread.java:748) ~[na:1.8.0_172]
Caused by: com.iota.iri.service.milestone.MilestoneException: unexpected error while analyzing the transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:239) ~[iri-1.7.0.jar:na]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:187) ~[iri-1.7.0.jar:na]
        ... 10 common frames omitted
Caused by: com.iota.iri.service.milestone.MilestoneException: error while checking milestone status of transaction XFRTRIJGPEAKJHZKKLLACOOGCVGBCTEGYFOVBKPAXCUYFRQQQEFVBOFYWLEHVQUYOXBDIG99EULS99999
        at com.iota.iri.service.milestone.impl.MilestoneServiceImpl.validateMilestone(MilestoneServiceImpl.java:245) ~[iri-1.7.0.jar:na]
        at com.iota.iri.service.milestone.impl.LatestMilestoneTrackerImpl.processMilestoneCandidate(LatestMilestoneTrackerImpl.java:212) ~[iri-1.7.0.jar:na]
        ... 11 common frames omitted
Caused by: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
        at java.util.LinkedList.checkElementIndex(LinkedList.java:555) ~[na:1.8.0_172]
        at java.util.LinkedList.get(LinkedList.java:476) ~[na:1.8.0_172]
        at com.iota.iri.service.milestone.impl.MilestoneServiceImpl.validateMilestone(MilestoneServiceImpl.java:195) ~[iri-1.7.0.jar:na]
        ... 12 common frames omitted
```
As soon as this tx is in the db this exception happens when validating milestones (because the bundle has only two outgoing transactions).

**Comment by georgmittendorfer** [comment]
This is fixed in 1.7.1 from my point of view.
